### ISSUE TITLE
left- and right-extension fails when alignment score turns negative

### ISSUE THREAD

**Issue by joellindegger** [Tongdongq/darwin-gpu#2]
### Expected behavior
For every candidate pair passed to the GPU implementation of GACT, reads should be either
1. filtered out after the first tile due to bad scores, or
2. aligned fully

### Observed behavior
Candidate pairs which are not filtered in the first tile are **not** aligned fully.

### Steps to reproduce
1. disable the first tile heuristic by setting `first_tile_score_threshold` in _params.cfg_ very low, e.g. -1000
2. track left-extension of candidate pairs, e.g. insert the following code at _gact.cpp:316_
    `printf("terminated callidx=%d reverse at query_pos=%d terminate=%d\n", callidx, query_pos, terminate[t]);`
3. observe that left-extension of most candidate pairs is terminated before the entire query is covered, such as
    `terminated callidx=168534 reverse at query_pos=1632 terminate=1`

We observe analogous behaviour for the right-extension.

### Underlying cause and proposed solution
The likely source of this behavior is _cuda_header.h:176_ and the following lines, which implement a maximum operation of several values.
The initialization `int tmp2 = 0;` incorrectly limits tmp2 to positive values: Smith-Waterman alignment _should_ be able to produce negative scores.

A possible solution is the initialization `int tmp2 = 1<<31; //min int`
Thus the following lines will correctly handle negative alignment scores.
With this change in place, we observe candidate pairs are fully left extended, as desired.
I.e. with above print statements, we see
`terminated callidx=168534 reverse at query_pos=0 terminate=0`

**Comment by Tongdongq** [comment]
Shouldn't Smith-Waterman set the negative scores to 0 automatically?
I think Needleman-Wunsch allows negative scores.

**Comment by joellindegger** [comment]
Thanks for the swift reply!

Yes, I think for *local* alignment, setting negative scores to 0 would be correct.
For *semi-global* or *global* alignment, negative scores should be possible.

To maintain coherence between tiles, the next tile should start where the previous tile ended (otherwise parts of the query are deleted without cost). However, local alignment does *not* enforce the alignment to start at the beginning of the tile.

For left-extension (right-extension), I think it requires top-right (bottom-left) pinned semi-global alignment.
In this case, negative scores **should** be considered.

**Comment by Tongdongq** [comment]
That makes sense, but the [original paper](https://users.cs.duke.edu/~lkw34/papers/darwin-asplos2018.pdf) describes that the modified Smith-Waterman starts traceback in the corner (except for the first tile).
I didn't create the original CPU implementation which can be found [here](https://github.com/yatisht/darwin).
Perhaps you can find details of this traceback more easily in there.

**Comment by joellindegger** [comment]
I agree on these points, perhaps local alignment can indeed make sense under these restrictions.

However, from the algorithm description in the original Darwin paper, it is my understanding that alignment of the entire candidate location should **not** be aborted simply because negative scores _are_ or _would be_ encountered.
Is that correct?

This seems to be the current behavior of Darwin-GPU according to above methodology.

**Comment by Tongdongq** [comment]
Where did you find that point?
Do you know how the CPU version behaves in cases like these? Perhaps @yatisht can help explain the algorithm.
